gradio/utils.py

```diff
@@ -1040,3 +1040,49 @@ def __setitem__(self, key: K, value: V) -> None:
 
 def get_cache_folder() -> Path:
     return Path(os.environ.get("GRADIO_EXAMPLES_CACHE", "gradio_cached_examples"))
+
+
+def diff(old, new):
+    def compare_objects(obj1, obj2, path=None):
+        if path is None:
+            path = []
+        edits = []
+
+        if obj1 == obj2:
+            return edits
```

Comment:
could put this after the type check since this will check every entry which is more expensive than the typecheck operation

Reply:
I don't think it makes a difference. If they are different types, e.g. a list and a dict, it will fail this check quickly and go on. If they are the same type, e.g. two lists, yes it will go through every entry, but it has to do that anyway.

If you flip the order: with a list and a dict, it will do the type check first and return quickly. But with two lists, first it'll do an extra type check, then it will do equality check, and go through all the items anyway. So it's the same amount of checks.